### PR TITLE
Limit buildpod label length

### PR DESCRIPTION
**Pull request by csrwng**
Pod name and build name allow for up to 256 chars. However, a label value cannot be more than 64 chars. This change truncates the value of the pod label to 64 chars and relies on the build annotation to get the full name. 

BZ 1327500

### CODE REVIEW

**Comment by csrwng** [comment]
@bparees ptal

btw, I thought I saw some place where pods were getting selected by label value, but there isn't. All we select on is the existence of the label.

File: pkg/build/api/util_test.go [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+	"testing"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+)
+
+func TestGetBuildPodName(t *testing.T) {
+	if expected, actual := "mybuild-build", GetBuildPodName(&Build{ObjectMeta: kapi.ObjectMeta{Name: "mybuild"}}); expected != actual {
+		t.Errorf("Expected %s, got %s", expected, actual)
+	}
+}
```
Comment (bparees) [inline pkg/build/api/util_test.go]: test the truncator?

File: pkg/build/controller/strategy/custom_test.go [hunk 2]
```diff
@@ -150,6 +150,21 @@ func TestCustomCreateBuildPodWithCustomCodec(t *testing.T) {
...
+	build := mockCustomBuild(false, false)
+	build.Name = strings.Repeat("a", validation.DNS1123LabelMaxLength*2)
+	pod, err := strategy.CreateBuildPod(build)
+	if err != nil {
+		t.Fatalf("unexpected: %v", err)
+	}
+	if pod.Labels[buildapi.BuildLabel] != build.Name[:validation.DNS1123LabelMaxLength] {
+		t.Errorf("Unexpected build label value: %s", pod.Labels[buildapi.BuildLabel])
+	}
+}
```
Comment (bparees) [inline pkg/build/controller/strategy/custom_test.go]: ignore my earlier comment, i see you're testing truncation here.

**Comment by bparees** [comment]
@csrwng so the build/reaper/reaper.go is doing selection based on the BuildConfigLabel, so aren't we going to have the same problem w/ that label?

**Comment by bparees** [comment]
(that is a label on Build objects.  Not sure if it's also on Pods or not).

**Comment by csrwng** [comment]
Yes, you are right, we would have the same problem with labels on builds. I'll fix that as well.

**Comment by csrwng** [comment]
@bparees updated to truncate buildconfig label on builds

File: pkg/build/api/helpers.go [hunk 3]
```diff
@@ -41,7 +41,8 @@ func FilterBuilds(builds []Build, predicate PredicateFunc) []Build {
 // ByBuildConfigLabelPredicate matches all builds that have build config label with specified value.
 func ByBuildConfigLabelPredicate(labelValue string) PredicateFunc {
 	return func(arg interface{}) bool {
-		return (hasBuildConfigLabel(arg.(Build), BuildConfigLabel, labelValue) ||
+		return (hasBuildConfigAnnotation(arg.(Build), BuildConfigAnnotation, labelValue) ||
+			hasBuildConfigLabel(arg.(Build), BuildConfigLabel, labelValue) ||
 			hasBuildConfigLabel(arg.(Build), BuildConfigLabelDeprecated, labelValue))
 	}
 }
```
Comment (bparees) [inline pkg/build/api/helpers.go]: this function should be renamed now and the godoc updated.

File: pkg/build/api/helpers.go [hunk 4]
```diff
@@ -50,3 +51,11 @@ func hasBuildConfigLabel(build Build, labelName, labelValue string) bool {
 	value, ok := build.Labels[labelName]
 	return ok && value == labelValue
 }
+
+func hasBuildConfigAnnotation(build Build, labelName, labelValue string) bool {
```
Comment (bparees) [inline pkg/build/api/helpers.go]: annotationName, annotationValue

File: pkg/build/api/util.go [hunk 5]
```diff
@@ -45,3 +50,25 @@ func SourceType(source BuildSource) string {
...
+// given Pod.
+func GetBuildName(pod *kapi.Pod) string {
+	return pod.Annotations[BuildAnnotation]
+}
+
+// GetBuildConfigName returns the name of a BuildConfig associated
+// with the given Build.
+func GetBuildConfigName(build *Build) string {
+	return build.Annotations[BuildConfigAnnotation]
+}
```
Comment (bparees) [inline pkg/build/api/util.go]: we need to handle existing builds/buildconfigs that won't have this annotation, so if this is empty i think you need to fallback to the label value, right?

Comment (bparees) [inline pkg/build/api/util.go]: (same thing for GetBuildName, unless we were already always adding the BuildAnnotation to pods?)

Comment (bparees) [inline pkg/build/api/util.go]: looks like you're not actually calling GetBuildConfigName anywhere?

Comment (csrwng) [inline pkg/build/api/util.go]: yeah, I need to remove it. We do have a function elsewhere to get the build config name that does take the old labels into consideration.

**Comment by csrwng** [comment]
pushed fixes

File: pkg/build/api/util.go (hunk 5 above)
Comment (csrwng) [inline pkg/build/api/util.go]: we were already adding BuildAnnotation to pods

**Comment by bparees** [comment]
squash and lgtm

**Comment by csrwng** [comment]
[testonlyextended][extended:core(builds)]

**Comment by csrwng** [comment]
Extended test failure is xip.io flake
[merge]

**Pull request merged**